### ISSUE TITLE
Eventum outgoing email content type issue

### ISSUE THREAD

**Issue by eventumrcf** [eventum/eventum#982]
**Version(s) affected**: 3.9.6

**PHP Version**: 7.3.23

**Browser and Version**: Chrome 87.0.4280

**Description**  

Hi , after upgrading from 2.3.6 (I followed the complete upgrade path), my outgoing e-mails that have attachments seem to all show as text, for example  a PDF attached to an email would look like this:

The From field has strange characters in it: 
```
From: =?UTF-8?Q?My=20Name?=
```

The headers look like this:

```
This is a message in Mime Format.  If you see this, your mail reader does not support this format.

--=_fe716eb07a166df157f3c0826a6c376b
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is a message in Mime Format.  If you see this, your mail reader does not support this format.

--=_4a063a24d45eb4a96a1cc7e94d45dcfc
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

test9
--=_4a063a24d45eb4a96a1cc7e94d45dcfc
Content-Type: application/pdf; charset=UTF-8
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="Estimate EST000636.pdf"

Then the PDF code starts:
%PDF-1.4
```

Pure HTML e-mails come into the system as base64 encoded:

```
EdDvITEwaaNwJKAHFgv3jf/hGo0fNKYHpNo8/DSEoaTCpXu+mLXtOVJASvrTl9h7/qQaB6EBMCG6+/L9Xkzy2WyABW3vBaur3UJXYY9oOkaTBhrWMXp3ckm2QIJoZXpFiaf7qzhGP7Ty6D0hy8Y6UkF5SRFq8EeYFpn8djB0GZUYb7eLh+kOZDkzu4Em/yHpkElurvIX7G3si+1sbb+gZJ0KzNYhXf7seg+LT8O8Wn+SETNY3TcgUiLXeb7BqqDU/++Pm+Xu5nnWhgYpNXxldRAnSNqGzm0mKsTAUekeN5QHYagkOvasJhGXfVCRDtGxoLHfRXhRcSn+u0BAoyl40W7JLx4TFVSsIEpeyhf2nNp1EdwHmVi6pY/8S6ghe/vfStc8IMObauI85bH3VCZdXXo8J0J7BdcKo9+8/VEPXSZqGsKm1FiQN5mmrQK1DRnH5eMlIDdarI8HUFHo/ax2qPbIcHJDDfHK/EkvaUp7A9AHKaNlm2C4V+q49zrU98sNQTnw0j7XAZUqQCgnDdKxJ2sfuVwHhprVXntVZyt0AJ5RQjde+0FicECPzq8pvue/cl1xI4dpufB5wtj8mLV8IiiyCBUEkx7FeIjxUA=
Content-Type: text/html;
 charset="utf-8"
Content-ID: <F30ED3A4D6B9EA449314B56B54853B48@eurprd05.prod.outlook.com>
Content-Transfer-Encoding: base64
X-IMAP-UnixDate: 1608463558
To: Target User<targetuser@gmail.com>
X-Eventum-Type: other_email
X-Eventum-Assignee: 
X-Eventum-Category: Test
X-Eventum-Project: Sales
X-Eventum-Priority: Immediate
X-Eventum-CustomField-Company: Company name
X-Eventum-CustomField-Email: fromuser@somesite.com
X-Eventum-CustomField-First-Name: test
X-Eventum-CustomField-Phone-Number: 021384484848
X-Eventum-CustomField-Country: London
X-Eventum-CustomField-Conversion: Yes
X-Eventum-CustomField-Type-Of-Business: TEst
Precedence: bulk
Auto-Submitted: auto-generated

PGh0bWw+DQo8aGVhZD4NCjxtZXRhIGh0dHAtZXF1aXY9IkNvbnRlbnQtVHlwZSIgY29udGVudD0i
dGV4dC9odG1sOyBjaGFyc2V0PXV0Zi04Ij4NCjwvaGVhZD4NCjxib2R5Pg0KPGRpdiBkaXI9Imx0
ciI+RGVhciBNYXJ0aW4gOg0KPGRpdj5UaGUgYXR0YWNobWVudCBpcyA3IGJsYW5rIHBhZ2VzJm5i
```

Any ideas would be appreciated
Regards
RCF

**How to reproduce**  

1) Send a message from eventum with an attachment
2) Send a HTML e-mail to an eventum project

**Possible Solution**  

**Additional context**  

It seems to me that it is not doing the whole content-type multipart bit for some reason, it looks like it is just sending as text every time.

EDIT: I actually see the correct header further up in the message so it is there:
```
MIME-Version: 1.0
Content-Type: multipart/mixed;
 boundary="=_20adbf278c395e5afde1ad3a49d82896"
```

But all clients including google displays the email including the binary code of the pdf as inline text.

**Comment by glensc** [comment]
This bug report is difficult to grasp. Please edit and reformat your bug report with markdown. Use code-fence for example blocks.

Markdown guide:
- https://guides.github.com/features/mastering-markdown/

Also, include a copy of the email to reproduce as an attachment.

**Comment by cpinfold** [comment]
I had a similar (possibly the same) issue on version 3.8.7 (I still need to upgrade)

Note that the issue started happening out of the blue and I had not upgraded versions.

I was able to track it down to the X-Ms-Exchange-Antispam-Messagedata header added by Office365/ExchangeOnline to messages.  It never happened to my messages sent by other providers.  I don't know if the header is incorrectly formatted by Microsoft or Eventum is interpreting it badly.  

I had hoped to address it through the workflow API, but `handleNewEmail` is called after the email is processed and I don't believe that there is a method called before the email is processed.  If there is one that anyone knows of, please let me know!

In `lib/eventum/class.support.php` I added the last line below to the beginning of the insertEmail method.
````
public static function insertEmail(MailMessage $mail, $email_options): int  
{                                                                           
    // Issue cause when this header is added to notifications (from Office365)
    $mail->getHeaders()->removeHeader('X-Ms-Exchange-Antispam-Messagedata');
````
While a terrible **kludge**, give it a try and advise if that works, that may point others to why it is happening.

I will send a followup and provide the header that caused the issue and an example before it started of one that didn't

**Comment by cpinfold** [comment]
Here is a bad header (incoming to Eventum):  

````
X-Ms-Exchange-Antispam-Messagedata: LjnjJdpGGk+cTVBUQzDxEQ5PK4r7tGGbhZZ/zYsmW3X6dcn+/bxvLzgL3kbCPvj9oayzcDs/xeUTn/iILPY2XsD8vZBL4F3h6/dev7yvn6AqoLaZdyUIxYfYWVHQ67jhdVrLhhVDOY2K8V9AV2NMvwc7jytA3ygU1dlMmmuIG+L79NIX8y6LMldSaUmBcaPRBIdcmAoVHg+TMMzQLG+SuL56f4Te06w6aonCfdcTQLjGxGOcFUaDOBqeCL/5zcejW/vsJnNmXFwRpzIGz3rmLAg/FeE5Rsl4DTuYEm1Paq6975Q7NTy2bUitnnGfqtXARsMFnDODlkgA7G5GljnZc+o2njG5rMj7IEhMDSOIbo/c6/zh2/trSMttBlW1AZzBWwl8xWH5GycTRPIw+xb1r6AS5uKb6QIrPgbd0FB9QyrBynyZu8pF30FSSQ6jenRFyOJ71sAwn42ohcdsh1UZ4xjdSeRzbGQDwGS8Vo8haOOl+PY6OR25oMVfe1pgaF4QnN7dxnIw3DU182F4Sa6HrkGZfi78TUI/hhWEqF5UvnCrnG0Khc6siugveKfJOnc1yAdJNhO0dkDsj6NEHf/Y5IfuKibowmklSTgjNcJ58LZ8oX8PSEzCFnp3Ky8yOzXf3zhbaZ8eYMecWjGY5xa2T+BaxW1bGeomh7fgBV5X5Uz/ZRCtSQEzIORJNnN+Bvl5c7fxbTXr6mg6zojn/XFO7x8OiYumCwzWGwzHe5sQDEKFqvsm4g36IDUXkGwhk178oUK062Hy7Hg1do25U69BJib/hQLQCVrh2lOuVJsL6005Po+V/BXA9GyJhHN1WRvYka5QZN9Iu3fv7XMvIPzqphoJaeFFQxabXHbQ+tGPm+hSi22MQkwG+M9kdvHcAH4fnk9DjoIfPZI4OHP7fttFw4vCpHgzSV4z/65IXtY+t7cgPa9ablO+d623P7pq4TIvpVtc7DSkkYVdGiRBZa14tapLEJPToX+QafT9o0OokwE=
````

That causes Eventum to send an email with the message body starting with:
`7DSkkYVdGiRBZa14tapLEJPToX+QafT9o0OokwE=`
and show the rest of the headers.

Here is part of what Eventum sends out based on the above being ingested:
````
...
X-Ms-Exchange-Antispam-Messagedata: LjnjJdpGGk+cTVBUQzDxEQ5PK4r7tGGbhZZ/zYsmW3X6dcn+/bxvLzgL3kbCPvj9oayzcDs/xeUTn/iILPY2XsD8vZBL4F3h6/dev7yvn6AqoLaZdyUIxYfYWVHQ67jhdVrLhhVDOY2K8V9AV2NMvwc7jytA3ygU1dlMmmuIG+L79NIX8y6LMldSaUmBcaPRBIdcmAoVHg+TMMzQLG+SuL56f4Te06w6aonCfdcTQLjGxGOcFUaDOBqeCL/5zcejW/vsJnNmXFwRpzIGz3rmLAg/FeE5Rsl4DTuYEm1Paq6975Q7NTy2bUitnnGfqtXARsMFnDODlkgA7G5GljnZc+o2njG5rMj7IEhMDSOIbo/c6/zh2/trSMttBlW1AZzBWwl8xWH5GycTRPIw+xb1r6AS5uKb6QIrPgbd0FB9QyrBynyZu8pF30FSSQ6jenRFyOJ71sAwn42ohcdsh1UZ4xjdSeRzbGQDwGS8Vo8haOOl+PY6OR25oMVfe1pgaF4QnN7dxnIw3DU182F4Sa6HrkGZfi78TUI/hhWEqF5UvnCrnG0Khc6siugveKfJOnc1yAdJNhO0dkDsj6NEHf/Y5IfuKibowmklSTgjNcJ58LZ8oX8PSEzCFnp3Ky8yOzXf3zhbaZ8eYMecWjGY5xa2T+BaxW1bGeomh7fgBV5X5Uz/ZRCtSQEzIORJNnN+Bvl5c7fxbTXr6mg6zojn/XFO7x8OiYumCwzWGwzHe5sQDEKFqvsm4g36IDUXkGwhk178oUK062Hy7Hg1do25U69BJib/hQLQCVrh2lOuVJsL6005Po+V/BXA9GyJhHN1WRvYka5QZN9Iu3fv7XMvIPzqphoJaeFFQxabXHbQ+tGPm+hSi22MQkwG+M9kdvHcAH4fnk9DjoIfPZI4OHP7fttFw4vCpHgzSV4z/65IXtY+t7cgPa9ablO+d623P7pq4TIvpV tc

7DSkkYVdGiRBZa14tapLEJPToX+QafT9o0OokwE=
```
**Note:** The added space and line breaks.  The mail program interprets the double newline as the start of the message body.

**Comment by eventumrcf** [comment]
Thanks so much for the information, I will explore this a bit today and let you know if I made any progress with regards to this.

**Comment by eventumrcf** [comment]
Will also attach sample email as requested.

**Comment by eventumrcf** [comment]
OK the incoming e-mails seem to be sorted by removing that header. 

I've been spending time trying to figure out what the issue is with the emails that are not sending correctly from within the issues themselves, I have tracked it down to when you use the CC functionality for someone that is not added to the notification/authorised replier list, i.e. when it uses this function in class.support.php:

````
self::sendDirectEmail($mail, $options);
````

This is only applicable when adding attachments. 
When sending mail via this function:

````
 $sup_id = self::insertEmail($mail, $email_options);
````
Everything works and the attachments display as they should.

I am attaching a sample e-mail with a sample PDF attached.


[Testing CC field.zip](https://github.com/eventum/eventum/files/5787471/Testing.CC.field.zip)

**Comment by eventumrcf** [comment]
Another update, I think I found the problem - I made a text attachment:

````
Content-Type: multipart/mixed;
 boundary="=_cf8f7b6d3246fb1a04fc5e81f08a2a4a"
Precedence: bulk
Auto-Submitted: auto-generated


This is a message in Mime Format.  If you see this, your mail reader does not support this format.

--=_cf8f7b6d3246fb1a04fc5e81f08a2a4a
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This is a message in Mime Format.  If you see this, your mail reader does not support this format.

--=_2fb4cdc2f910a21ec7eb4a0a52d3f42a
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Test text attachment
--=_2fb4cdc2f910a21ec7eb4a0a52d3f42a
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="textattachment.txt"

TESTING
TESTING
TESTING
TESTING
TESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTING

--=_2fb4cdc2f910a21ec7eb4a0a52d3f42a--
--=_cf8f7b6d3246fb1a04fc5e81f08a2a4a
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit
Content-Disposition: attachment; filename="textattachment.txt"

TESTING
TESTING
TESTING
TESTING
TESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTINGTESTING

--=_cf8f7b6d3246fb1a04fc5e81f08a2a4a--
````
As far as I understand it, the boundary is cf8f7b6d3246fb1a04fc5e81f08a2a4a, the attachment should not appear twice and with a different boundary the second time (2fb4cdc2f910a21ec7eb4a0a52d3f42a)?

**Comment by eventumrcf** [comment]
OK tracked it down to the following:

In \lib\eventum\class.support.php - in the function sendEmail:

````
$mail = self::buildMail($issue_id, $from, $to, $cc, $subject, $body, $in_reply_to, $iaf_ids);
````

There the mail object is constructed, i.e. the function toMailMessage() in \src\Mail\MailBuilder.php  is already being called.
When sendDirectEmail is called from \lib\eventum\class.support.php - the $mail object is passed, and inside that function the body is retrieved like so:

````
$body = $mail->getContent();
````

In the recipient loop:
$builder uses $body (which has already been processed) from 

````
$builder
                ->addTextPart($body)
                ->getMessage()
                ->setSubject($subject)
                ->setTo($recipient)
                ->setFrom($from);
````
And then calls 

````
$mail = $builder->toMailMessage(); 
````
again, which results in double-processing of the body.

I've fixed this by passing the unprocessed body through to sendDirectEmail, and using that in the recipient loop, which solves the mime boundary issues. 

I now have another issue though, binary attachments are corrupted it seems (A text attachment works fine) - looking into this.

**Comment by vladsf** [comment]
Hi, I guess I found the source of trouble with long X-Ms-Exchange-Antispam* headers and submitted PR [Fold long lines during smtp communication..](https://github.com/laminas/laminas-mail/pull/138#issue-580790129)

**Comment by vladsf** [comment]
a quick and dirty fix might be (while PR is being reviewed):
```
--- vendor/laminas/laminas-mail/src/Protocol/Smtp.php.1	2021-02-25 17:52:34.598451053 +0000
+++ vendor/laminas/laminas-mail/src/Protocol/Smtp.php	2021-02-25 17:52:58.277500727 +0000
@@ -328,7 +328,7 @@
         rewind($fp);
 
         // max line length is 998 char + \r\n = 1000
-        while (($line = stream_get_line($fp, 1000, "\n")) !== false) {
+        while (($line = stream_get_line($fp, 2000, "\n")) !== false) {
             $line = rtrim($line, "\r");
             if (isset($line[0]) && $line[0] === '.') {
                 // Escape lines prefixed with a '.'
```

**Comment by glensc** [comment]
btw (not related to the problem, but curiosity), tried to find info on what the actual `X-Ms-Exchange-Antispam-Messagedata` header contains. could not find an exact match, but only docs of similar headers:

- https://docs.microsoft.com/en-us/exchange/antispam-and-antimalware/antispam-protection/antispam-stamps?view=exchserver-2019
- https://docs.microsoft.com/en-us/exchange/antispam-and-antimalware/antispam-protection/view-antispam-stamps-in-outlook?view=exchserver-2019

my interest was to figure out does sharing them in public could leak some security details.

**Comment by vladsf** [comment]
No idea. X-Ms-Exchange-Antispam-Messagedata becomes 1000+ if an email is long and has an attachment. May be not the in the every case but I could test it by creating such emails. 

Also MS Exchange respect "MS" uppercase, and originally sends X-M**S**-Exchange* but passing through several servers it might become one-liner and X-M**s**-Exchange.

**Comment by glensc** [comment]
The long header had been fixed in 2.14 or láminas mail
- https://github.com/laminas/laminas-mail/pull/140

So this is no longer relevant.

Create a new issue if problem is not solved, as it's likely another angle that needs to be fixed
